mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -5,821 +5,745 @@
-            deserialize_from(&mut Cursor::new(slice))
-                .context(DeserializeError {})?;
-
-        if header.header_size != MetaDataHeader::METADATA_HEADER_SIZE {
-            return Err(MetaDataError::HeaderSize {
-                actual_size: header.header_size,
-                expected_size: MetaDataHeader::METADATA_HEADER_SIZE,
-            });
-        }
+            deserialize_from(&mut reader).context(DeserializeError {})?;
 
         if header.signature != [0x4d, 0x61, 0x79, 0x61, 0x44, 0x61, 0x74, 0x61]
         {
-            return Err(MetaDataError::HeaderSignature {});
+            return Err(MetaDataError::ObjectSignature {});
+        }
+
+        if header.count > MetaDataObject::MAX_CHILD_ENTRIES {
+            return Err(MetaDataError::ChildTableSize {});
+        }
+
+        let mut children: Vec<MetaDataChildEntry> =
+            Vec::with_capacity(header.count as usize);
+
+        for _ in 0 .. header.count {
+            children.push(
+                deserialize_from(&mut reader).context(DeserializeError {})?,
+            );
+        }
+
+        if header.table_crc != MetaDataChildEntry::checksum(&children) {
+            return Err(MetaDataError::ChildTableChecksum {});
         }
 
         let checksum = header.self_checksum;
 
         if header.checksum() != checksum {
-            return Err(MetaDataError::HeaderChecksum {});
+            return Err(MetaDataError::ObjectChecksum {});
         }
 
-        Ok(header)
+        let object = MetaDataObject {
+            signature: header.signature,
+            self_checksum: header.self_checksum,
+            generation: header.generation,
+            timestamp: header.timestamp,
+            table_crc: header.table_crc,
+            children,
+        };
+
+        Ok(object)
     }
 
-    /// Checksum the header with the checksum field itself set to 0
     pub fn checksum(&mut self) -> u32 {
-        self.self_checksum = 0;
-        self.self_checksum = crc32::checksum_ieee(&serialize(self).unwrap());
+        let mut header = MetaDataHeader::from(&*self);
+        self.self_checksum = header.checksum();
         self.self_checksum
     }
 
-    /// Generate a new MetaDataHeader based on the partition information and the
-    /// underlying block_size
-    pub fn new(block_size: u32, partition: &GptEntry) -> MetaDataHeader {
-        let index_start = Aligned::get_blocks(
-            MetaDataHeader::METADATA_HEADER_SIZE,
-            block_size,
-        );
-
-        let data_start = index_start
-            + Aligned::get_blocks(
-                MetaDataHeader::MAX_INDEX_ENTRIES
-                    * MetaDataHeader::INDEX_ENTRY_SIZE,
-                block_size,
-            );
+    /// Perform basic checks to ensure object is valid,
+    /// and then calculate checksums.
+    pub fn validate(&mut self) -> Result<(), MetaDataError> {
+        if self.signature != [0x4d, 0x61, 0x79, 0x61, 0x44, 0x61, 0x74, 0x61] {
```

Comment:
It strikes me that this is a magic value which we should probably const so we could use it and find it later.

Reply:
Done